app/code/Magento/Catalog/Model/MaskedListIdToCompareListId.php

```diff
@@ -0,0 +1,53 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\Catalog\Model;
+
+use Magento\Catalog\Model\ResourceModel\Product\Compare\ListIdMask as ListIdMaskMaskResource;
+
+/**
+ * MaskedListId to ListId resolver
+ */
+class MaskedListIdToCompareListId
+{
+    /**
+     * @var ListIdMaskFactory
+     */
+    private $listIdMaskFactory;
+
+    /**
+     * @var ListIdMaskMaskResource
+     */
+    private $listIdMaskResource;
+
+    /**
+     * @param ListIdMaskFactory $listIdMaskFactory
+     * @param ListIdMaskMaskResource $listIdMaskResource
+     */
+    public function __construct(
+        ListIdMaskFactory $listIdMaskFactory,
+        ListIdMaskMaskResource $listIdMaskResource
+    ) {
+        $this->listIdMaskFactory = $listIdMaskFactory;
+        $this->listIdMaskResource = $listIdMaskResource;
+    }
+
+    /**
+     * Get maskedId by listId
+     *
+     * @param string $maskedListId
+     *
+     * @return int
+     */
+    public function execute(string $maskedListId): int
+    {
+        $listIdMask = $this->listIdMaskFactory->create();
+        $this->listIdMaskResource->load($listIdMask, $maskedListId, 'masked_id');
+
+        return (int)$listIdMask->getListId();
```

Comment:
Have you tested the scenario when compare list ID is missing in the DB for the provided masked value?

Reply:
Hi @rogyar 
We will get `0` in this case 
<img width="1423" alt="Screenshot 2020-11-04 at 17 55 21" src="https://user-images.githubusercontent.com/13517181/98134077-09ff8d00-1ec7-11eb-80ac-36d8e5795b6e.png">